// @siyuan-community/zhi-common/src/lib/md-adaptor/ShowdownAdaptor.js
/*
 * Copyright (c) 2023, Terwer . All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Terwer designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Terwer in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Terwer, Shenzhen, Guangdong, China, youweics@163.com
 * or visit www.terwer.space if you need additional information or have any
 * questions.
 */
System.register(["showdown", "../utils/zhi-common-util.js"], function (exports_1, context_1) {
    "use strict";
    var showdown_1, zhi_common_util_js_1, ShowdownAdaptor;
    var __moduleName = context_1 && context_1.id;
    return {
        setters: [
            function (showdown_1_1) {
                showdown_1 = showdown_1_1;
            },
            function (zhi_common_util_js_1_1) {
                zhi_common_util_js_1 = zhi_common_util_js_1_1;
            }
        ],
        execute: function () {
            /**
             * showdown 适配器
             *
             * @author terwer
             * @version 1.0.0
             * @since 1.0.0
             */
            ShowdownAdaptor = class ShowdownAdaptor {
                logger;
                converter;
                constructor() {
                    this.logger = zhi_common_util_js_1.createCommonLogger("showdown-adaptor");
                    this.converter = new showdown_1.default.Converter();
                }
                isAvailable() {
                    return typeof showdown_1.default !== "undefined";
                }
                renderMarkdownStr(md) {
                    if (!this.isAvailable()) {
                        throw new Error("Showdown is not available");
                    }
                    this.logger.info("Showdown is rendering md to HTML...");
                    return Promise.resolve(this.converter.makeHtml(md));
                }
            };
            exports_1("default", ShowdownAdaptor);
        }
    };
});
