// @siyuan-community/zhi-common/src/lib/zhi-lib-common.js
System.register(["./dateUtil.js", "./strUtil.js", "./versionUtil.js", "./htmlUtil.js", "./markdownUtil.js", "./jsonUtil.js", "./objectUtil.js"], function (exports_1, context_1) {
    "use strict";
    var dateUtil_js_1, strUtil_js_1, versionUtil_js_1, htmlUtil_js_1, markdownUtil_js_1, jsonUtil_js_1, objectUtil_js_1, ZhiCommon;
    var __moduleName = context_1 && context_1.id;
    return {
        setters: [
            function (dateUtil_js_1_1) {
                dateUtil_js_1 = dateUtil_js_1_1;
            },
            function (strUtil_js_1_1) {
                strUtil_js_1 = strUtil_js_1_1;
            },
            function (versionUtil_js_1_1) {
                versionUtil_js_1 = versionUtil_js_1_1;
            },
            function (htmlUtil_js_1_1) {
                htmlUtil_js_1 = htmlUtil_js_1_1;
            },
            function (markdownUtil_js_1_1) {
                markdownUtil_js_1 = markdownUtil_js_1_1;
            },
            function (jsonUtil_js_1_1) {
                jsonUtil_js_1 = jsonUtil_js_1_1;
            },
            function (objectUtil_js_1_1) {
                objectUtil_js_1 = objectUtil_js_1_1;
            }
        ],
        execute: function () {
            exports_1("DateUtil", dateUtil_js_1.default);
            exports_1("StrUtil", strUtil_js_1.default);
            exports_1("VersionUtil", versionUtil_js_1.default);
            exports_1("HtmlUtil", htmlUtil_js_1.default);
            exports_1("MarkdownUtil", markdownUtil_js_1.default);
            exports_1("JsonUtil", jsonUtil_js_1.default);
            exports_1("ObjectUtil", objectUtil_js_1.default);
            /**
             * 平台无关的通用工具类
             *
             * @author terwer
             * @version 1.4.0
             * @since 1.3.0
             */
            ZhiCommon = class ZhiCommon {
                dateUtil;
                strUtil;
                versionUtil;
                htmlUtil;
                markdownUtil;
                jsonUtil;
                objectUtil;
                constructor() {
                    this.dateUtil = new dateUtil_js_1.default();
                    this.strUtil = new strUtil_js_1.default();
                    this.versionUtil = new versionUtil_js_1.default();
                    this.htmlUtil = new htmlUtil_js_1.default();
                    this.markdownUtil = new markdownUtil_js_1.default();
                    this.jsonUtil = new jsonUtil_js_1.default();
                    this.objectUtil = new objectUtil_js_1.default();
                }
            };
            exports_1("ZhiCommon", ZhiCommon);
        }
    };
});
